src/query/storage/m3/consolidators/multi_iterator.go

```diff
@@ -0,0 +1,285 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package consolidators
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/ts"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+type data struct {
+	dp         ts.Datapoint
+	unit       xtime.Unit
+	annotation ts.Annotation
+}
+
+type idxData struct {
+	data
+	isBase bool
+	idx    int
+}
+
+func buildBaseIdxData(it encoding.SeriesIterator) idxData {
+	dp, unit, annotation := it.Current()
+	return idxData{
+		isBase: true,
+		data: data{
+			dp:         dp,
+			unit:       unit,
+			annotation: annotation,
+		},
+	}
+}
+
+func buildIndexIdxData(it encoding.SeriesIterator, idx int) idxData {
+	dp, unit, annotation := it.Current()
+	return idxData{
+		idx: idx,
+		data: data{
+			dp:         dp,
+			unit:       unit,
+			annotation: annotation,
+		},
+	}
+}
+
+type idxDatas []idxData
+
+func (d idxDatas) insert(insertData idxData) {
+	addingTime := insertData.data.dp.Timestamp
+	idx := len(d)
+	for i, data := range d {
+		existingTime := data.dp.Timestamp
+		if existingTime.Before(addingTime) {
+			continue
+		}
+
+		if existingTime.Equal(addingTime) {
+			if data.isBase || data.idx < insertData.idx {
+				continue
+			}
+		}
+
+		idx = i
+		break
+	}
+
+	d = append(d, idxData{})
+	copy(d[:idx+1], d[:idx])
+	d[idx] = insertData
+
+	d = append(d[:idx], insertData)
+	d = append(d, d[:idx]...)
+}
+
+// NB: enforce multiIterator implements encoding.SeriesIterator.
+var _ encoding.SeriesIterator = (*multiIterator)(nil)
+
+type multiIterator struct {
+	err error
+
+	current   data
+	indexData idxDatas
+	iters     []encoding.SeriesIterator
+	base      encoding.SeriesIterator
+}
+
+func (it *multiIterator) Next() bool {
+	if len(it.indexData) == 0 {
+		return false
+	}
+
+	currTs := it.current.dp.Timestamp
+	for {
+		if len(it.indexData) == 0 {
+			return false
+		}
+
+		// NB: take off value corresponding to next ordered iterator.
+		first := it.indexData[0]
+		copy(it.indexData, it.indexData[1:])
```

Comment:
Interesting, was it not worth while seeing if we could use the already stack based "iterators.go"?

It has the whole concept of "values" and "earliest" and applies the different timestamp de-duping policies. 

Is there a blocker for using it as the parent de-duplicater?
https://github.com/m3db/m3/blob/master/src/dbnode/encoding/iterators.go